drawcpu.1
.TH DRAWCPU 1

.SH NAME
drawcpu  \- connection to cpu, fs, and auth servers

.SH SYNOPSIS
.B drawcpu
[
.B -9GBO
] [
.B -h
.I host
] [
.B -u
.I user
] [
.B -a
.I authserver
] [
.B -s
.I secstore
] [
.B -e
\fR'\fIcrypt hash\fR'
] [
.B -k
.I keypattern
] [
.B -p
] [
.B -t
.I timeout
] [
.B -r
.I root
] [
.B -g
.I geometry
] [
.B -c
.I cmd \fR...]

.SH DESCRIPTION
.I Drawterm
is a client for connecting venerable systems to Plan 9 systems. The standard behavior with no options provided is to begin a graphical session which will prompt for a cpu server, auth server, and password. If $USER is not set, drawcpu will prompt for a username as well. 

The goal of drawcpu is to provide an abstraction layer from the client operating system to the desired Plan 9 system. Client-side devices that can be represented as filesystems will be and are served from the namespace drawcpu operates in. Generally, this means that drawcpu behaves like a Plan 9 kernel and will attempt to reconstruct a Plan 9 terminal-like experience from a non-Plan 9 system.

.PP
The options are:
.PD

.TP
.B -9
Start drawcpu in 9P mode;
exporting the namespace over 9P on standard input and output.
Graphics and keyboard input are disabled.

.TP
.B -G
Start drawcpu in text-only mode (no graphics).

.TP
.B -B
Disable kbdfs.

.TP
.B -O
Use the old
.IR cpu (1)
to connect to the cpu server rather than 
.IR rcpu (1)\fR.

.TP
.B -h \fIhost
Connect to \fIhost\fR for cpu.

.TP
.B -u \fIuser
Remote user id.

.TP
.B -a \fIauthserver
Connect to \fIauthserver\fR for auth.

.TP
.B -s \fIsecstore
Sets the address of the
.IR secstore (8)
server. If the -s option is absent,
.IR secstore (1)
will attempt to dial tcp!$auth!secstore (or the otherwise specified auth server).

.TP
.B -e \fR'\fIcrypt hash\fR'
Specifies the \fR'\fIcrypt hash\fR'
for the connection. The default is 'rc4_256 sha1'. Usage of no encryption can be specified by 'clear' as per
.IR ssl (3)\fR. Note that this option is deprecated and only relevant to the outdated
.IR cpu (1)
protocol.

.TP
.B -k \fIkeypattern
Use keypattern to select a key to authenticate to the remote side (see 
.IR auth (2)\fR).

.TP
.B -p
Protect the connection against connection resets by establishing
.IR aan (8)
tunnel.

.TP
.B -t \fItimeout
Set the timeout for
.IR aan (8)
to a value in 
.I seconds\fR (default is one day).

.TP
.B -r \fIroot
Specifies the root directory on the client. The default is
.I /root
and all further paths are relative thereto.

.TP
.B -c \fIcmd \fR...
The command to run can be passed with -c cmd ..., otherwise an interactive shell is started. The user's profile is run before the command with $service set to cpu to allow further customization of the environment (see 
.IR rc (1)
for more information).

.PP
.SH ENVIRONMENT VARIABLES
.IP USER
Unless otherwise specified, the user stored in $USER is used for authentication.

.IP PASS
If $PASS is set, no password will be prompted for in authentication and the contents of $PASS will be used instead.

.IP cpu
Unless otherwise specified, the address stored in $cpu is used for the
.IR rcpu (1)
connection (if set).

.IP auth
Unless otherwise specified, the address stored in $auth is used for the 
.IR rcpu (1)
connection (if set).

.PP
.SH SERVICES
A number of services are provided in drawcpu. The exact functionality and availability of certain features may be dependent on your platform or architecture: 

.TP
.B /mnt/term
A mount of the client machine's root filesystem as well as certain virtual filesystems to present Plan 9 devices and interfaces that are not available on non-Plan 9 systems. On Windows this is a directory containing the available lettered disks (C:, A:, etc.). Additionally, there is always a /mnt/term/root folder which is a copy of the client machine's namespace with no virtual filesystems present as to avoid conflicting names (such as with Linux's /root).

.TP
.B #b
Assuming the -B flag is not set, /dev/kbd will be provided for kbdfs (see
.IR kbd (3)\fR).

.TP
.B #i
Assuming the -G flag is not set, various drawing device files will be provided in /dev (see
.IR draw (3)\fR).

.TP
.B #m
Assuming the -G flag is not set, files for controlling the mouse will provided in /dev (see
.IR mouse (3)\fR).

.TP
.B #c
A number of console device files giving access to the console screen and miscellaneous information are provided and mounted in /dev (see
.IR cons (3)\fR).

.TP
.B #I
The network filesystem is served and bound over /net, providing the interface to Internet Protocol stacks (see
.IR ip (3)\fR).

.TP
.B #A
An audio device filesystem is served, if possible, as a one-level directory in /dev (see
.IR audio (3)\fR).
Note that this device, if unable to be served, will not cause a panic in drawcpu.

.PP
.SH EXAMPLES
Make a headless rcpu session connecting to 10.30.167.25 using plan9.postnix.us as the auth server with
.IR aan (8)
enabled:
.IP
.EX
drawcpu -G -h 10.30.167.25 -a plan9.postnix.us -p
.EE
.PP

Make a session using cpu rather than rcpu to tenshi.postnix.us; this command is the same for connecting to a Plan 9 4th edition system:
.IP
.EX
drawcpu -O -h tenshi.postnix.us
.EE
.PP

.PP
.SH SOURCE
.B https://git.9front.org/plan9front/drawcpu/HEAD/info.html

.PP
.SH "SEE ALSO"
.IR rc (1),
.IR cpu (1),
.IR rcpu(1),
.IR con (1),
.IR import (4),
.IR exportfs (4),
.IR tlssrv (8),
.IR aan (8)

.PP
.SH BUGS
Drawterm is 
.I not
a Plan 9 program.

.PP
.SH HISTORY
Drawterm was originally developed by Russ Cox (rsc) for Plan 9 4th edition. This original version is still usable on Plan 9 and its forks which use the p9sk1 and older
.IR authsrv (6)
protocols.
.B https://swtch.com/drawcpu/

The 9front project has forked drawcpu to incorporate features from 9front, most importantly dp9ik authentication support (see 
.IR authsrv (6)\fR)
and the TLS-based
.IR rcpu (1)
protocol.


rc/drawcpu.c
/*
 * Plan 9 versions of system-specific functions
 *	By convention, exported routines herein have names beginning with an
 *	upper case letter.
 */
#include "rc.h"
#include "exec.h"
#include "io.h"
#include "fns.h"
#include "getflags.h"

static void execrfork(void);
static void execfinit(void);

builtin Builtin[] = {
	"cd",		execcd,
	"whatis",	execwhatis, 
	"eval",		execeval,
	"exec",		execexec,	/* but with popword first */
	"exit",		execexit,
	"shift",	execshift,
	"wait",		execwait,
	".",		execdot,
	"flag",		execflag,
	"finit",	execfinit,
	"rfork",	execrfork,
	"read",     execread,
	"ns",       execns,
	"bind",		execbind,
	"mount",	execmount,
	"unmount",	execunmount,
	"ls", 		execls,
	0
};

/* TODO: Set rcmain in .make  */
char Rcmain[]="/usr/local/lib/rcmain";
char Fdprefix[]="/fd/";

char *Signame[] = {
	"sigexit",	"sighup",	"sigint",	"sigquit",
	"sigalrm",	"sigkill",	"sigfpe",	"sigterm",
	0
};

static char *syssigname[] = {
	"exit",		/* can't happen */
	"hangup",
	"interrupt",
	"quit",		/* can't happen */
	"alarm",
	"kill",
	"sys: fp: ",
	"term",
	0
};

/*
 * finit could be removed but is kept for
 * backwards compatibility, see: rcmain.plan9
 */
static void
execfinit(void)
{
	char *cmds = estrdup("for(i in '/env/fn#'*){. -bq $i}\n");
	int line = runq->line;
	poplist();
	execcmds(openiocore(cmds, strlen(cmds)), estrdup(srcfile(runq)), runq->local, runq->redir);
	runq->lex->line = line;
	runq->lex->qflag = 1;
}

static void
execrfork(void)
{
	int arg;
	char *s;

	switch(count(runq->argv->words)){
	case 1:
		arg = RFENVG|RFNAMEG|RFNOTEG;
		break;
	case 2:
		arg = 0;
		for(s = runq->argv->words->next->word;*s;s++) switch(*s){
		default:
			goto Usage;
		case 'n':
			arg|=RFNAMEG;  break;
		case 'N':
			arg|=RFCNAMEG;
			break;
		case 'm':
			/*arg|=RFNOMNT;*/  break;
		case 'e':
			arg|=RFENVG;   break;
		case 'E':
			arg|=RFCENVG;  break;
		case 's':
			arg|=RFNOTEG;  break;
		case 'f':
			arg|=RFFDG;    break;
		case 'F':
			arg|=RFCFDG;   break;
		}
		break;
	default:
	Usage:
		pfmt(err, "Usage: %s [fnesFNEm]\n", runq->argv->words->word);
		setstatus("rfork usage");
		poplist();
		return;
	}
	if(rfork(arg)==-1){
		pfmt(err, "%s: %s failed\n", argv0, runq->argv->words->word);
		setstatus("rfork failed");
	} else {
		if(arg & RFCFDG){
			redir *rp;
			for(rp = runq->redir; rp; rp = rp->next)
				rp->type = 0;
		}
		setstatus("");
	}
	poplist();
}

char*
Env(char *name, int fn)
{
	static char buf[128];

	strcpy(buf, "/env/");
	if(fn) strcat(buf, "fn#");
	return strncat(buf, name, sizeof(buf)-1);
}

void
Vinit(void)
{
	int dir, fd, i, n;
	Dir *ent;

	dir = Open(Env("", 0), 0);
	if(dir<0){
		pfmt(err, "%s: can't open: %s\n", argv0, Errstr());
		return;
	}
	for(;;){
		ent = 0;
		n = dirread(dir, &ent);
		if(n <= 0)
			break;
		for(i = 0; i<n; i++){
			if(ent[i].length<=0 || strncmp(ent[i].name, "fn#", 3)==0)
				continue;
			if((fd = Open(Env(ent[i].name, 0), 0))>=0){
				io *f = openiofd(fd);
				word *w = 0, **wp = &w;
				char *s;
				while((s = rstr(f, "")) != 0){
					*wp = Newword(s, (word*)0);
					wp = &(*wp)->next;
				}
				closeio(f);
				setvar(ent[i].name, w);
				vlook(ent[i].name)->changed = 0;
			}
		}
		free(ent);
	}
	Close(dir);
}

char*
Errstr(void)
{
	static char err[ERRMAX];
	rerrstr(err, sizeof err);
	return err;
}

int
Waitfor(int pid)
{
	thread *p;
	Waitmsg *w;

	if(pid >= 0 && !havewaitpid(pid))
		return 0;

	while((w = wait()) != nil){
		delwaitpid(w->pid);
		if(w->pid==pid){
			setstatus(w->msg);
			free(w);
			return 0;
		}
		for(p = runq->ret;p;p = p->ret)
			if(p->pid==w->pid){
				p->pid=-1;
				p->status = estrdup(w->msg);
				break;
			}
		free(w);
	}

	if(strcmp(Errstr(), "interrupted")==0) return -1;
	return 0;
}

static void
addenv(var *v)
{
	word *w;
	int fd;
	io *f;

	if(v->changed){
		v->changed = 0;
		if((fd = Creat(Env(v->name, 0)))<0)
			pfmt(err, "%s: can't open: %s\n", argv0, Errstr());
		else{
			f = openiofd(fd);
			for(w = v->val;w;w = w->next){
				pstr(f, w->word);
				pchr(f, '\0');
			}
			flushio(f);
			closeio(f);
		}
	}
	if(v->fnchanged){
		v->fnchanged = 0;
		if((fd = Creat(Env(v->name, 1)))<0)
			pfmt(err, "%s: can't open: %s\n", argv0, Errstr());
		else{
			f = openiofd(fd);
			if(v->fn)
				pfmt(f, "fn %q %s\n", v->name, v->fn[v->pc-1].s);
			flushio(f);
			closeio(f);
		}
	}
}

static void
updenvlocal(var *v)
{
	if(v){
		updenvlocal(v->next);
		addenv(v);
	}
}

void
Updenv(void)
{
	var *v, **h;
	for(h = gvar;h!=&gvar[NVAR];h++)
		for(v=*h;v;v = v->next)
			addenv(v);
	if(runq)
		updenvlocal(runq->local);
	if(err)
		flushio(err);
}

void
Exec(char **argv)
{
	execvp(argv[0], argv+1);
}

int
Fork(void)
{
	Updenv();
	// TODO: Tie into the rendezvous so we can block 'em in rfork
	return rfork(RFPROC|RFFDG/*|RFREND*/);
}


typedef struct readdir readdir;
struct readdir {
	Dir	*dbuf;
	int	i, n;
	int	fd;
};

void*
Opendir(char *name)
{
	readdir *rd;
	int fd;
	if((fd = Open(name, 0))<0)
		return 0;
	rd = new(readdir);
	rd->dbuf = 0;
	rd->i = 0;
	rd->n = 0;
	rd->fd = fd;
	return rd;
}

static int
trimdirs(Dir *d, int nd)
{
	int r, w;

	for(r=w=0; r<nd; r++)
		if(d[r].mode&DMDIR)
			d[w++] = d[r];
	return w;
}

char*
Readdir(void *arg, int onlydirs)
{
	readdir *rd = arg;
	int n;
Again:
	if(rd->i>=rd->n){	/* read */
		free(rd->dbuf);
		rd->dbuf = 0;
		n = dirread(rd->fd, &rd->dbuf);
		if(n>0){
			if(onlydirs){
				n = trimdirs(rd->dbuf, n);
				if(n == 0)
					goto Again;
			}	
			rd->n = n;
		}else
			rd->n = 0;
		rd->i = 0;
	}
	if(rd->i>=rd->n)
		return 0;
	return rd->dbuf[rd->i++].name;
}

void
Closedir(void *arg)
{
	readdir *rd = arg;
	Close(rd->fd);
	free(rd->dbuf);
	free(rd);
}

static int interrupted = 0;

static void
notifyf(void* u, char *s)
{
	int i;

	USED(u);
	for(i = 0;syssigname[i];i++) if(strncmp(s, syssigname[i], strlen(syssigname[i]))==0){
		if(strncmp(s, "sys: ", 5)!=0) interrupted = 1;
		goto Out;
	}
	noted(NDFLT);
	return;
Out:
	if(strcmp(s, "interrupt")!=0 || trap[i]==0){
		trap[i]++;
		ntrap++;
	}
	noted(NCONT);
}

void
Trapinit(void)
{
	notify(notifyf);
}

long
Write(int fd, void *buf, long cnt)
{
	return write(fd, buf, cnt);
}

long
Read(int fd, void *buf, long cnt)
{
	return read(fd, buf, cnt);
}

long
Seek(int fd, long cnt, long whence)
{
	return seek(fd, cnt, whence);
}

int
Executable(char *file)
{
	Dir *statbuf;
	int ret;

	statbuf = dirstat(file);
	if(statbuf == nil)
		return 0;
	ret = ((statbuf->mode&0111)!=0 && (statbuf->mode&DMDIR)==0);
	free(statbuf);
	return ret;
}

int
Open(char *file, int mode)
{
	static int tab[] = {OREAD,OWRITE,ORDWR,OREAD|ORCLOSE};
	return open(file, tab[mode&3]);
}

void
Close(int fd)
{
	close(fd);
}

int
Creat(char *file)
{
	return create(file, OWRITE, 0666L);
}

int
Dup(int a, int b)
{
	return dup(a, b);
}

int
Dup1(int a)
{
	return dup(a, -1);
}

void
Exit(void)
{
	Updenv();
	exits(truestatus()?"":getstatus());
}

void
Noerror(void)
{
	interrupted = 0;
}

int
Isatty(int fd)
{
	return isatty(fd);
	/*
	char buf[64];
	if(fd2path(fd, buf, sizeof buf) != 0)
		return 0;
	return strlen(buf) >= 9 && strcmp(buf+strlen(buf)-9, "/dev/cons") == 0;
	*/
}

void
Abort(void)
{
	abort();
}

static int newwdir;

int
Chdir(char *dir)
{
	newwdir = 1;
	return chdir(dir);
}

void
Prompt(char *s)
{
	pstr(err, s);
	flushio(err);

	if(newwdir){
		char dir[4096];
		int fd;
		if((fd=Creat("/dev/wdir"))>=0){
			getwd(dir, sizeof(dir));
			Write(fd, dir, strlen(dir));
			Close(fd);
		}
		newwdir = 0;
	}
}
